pyfr/backends/opencl/kernels/errest.mako

```diff
@@ -0,0 +1,92 @@
+# -*- coding: utf-8 -*-
+<%inherit file='base'/>
+<%namespace module='pyfr.backends.base.makoutil' name='pyfr'/>
+
+__kernel void
+errest(int nrow, int ncolb, int ldim, __global fpdtype_t *__restrict__ err,
+       ${', '.join('__global const fpdtype_t* restrict {0}'.format(i)
+                   for i in ['x', 'y', 'z'])},
+       fpdtype_t atol, fpdtype_t rtol)
+
+{
+    __local fpdtype_t ${','.join('sdata{0}[{1}]'.format(i, sharesz)
+                                    for i in range(ncola))};
+    fpdtype_t r;
+
+    int idx, tstride;
+    int i = get_global_id(0), tid = get_local_id(0);
+    int bid = get_group_id(0), sharesz = get_local_size(0);
+    int gdim = (ncolb + sharesz - 1)/sharesz;
+    int lastblksize = ncolb % sharesz;
+
+    if (i < ncolb)
+    {
+        // All threads load the residuals of the first row to shared memory
+    % for k in range(ncola):
+        idx = SOA_IX(i, ${k}, ${ncola});
+        sdata${k}[tid] = pow(x[idx]/(atol + rtol*max(fabs(y[idx]),
+                                                     fabs(z[idx]))), 2);
+    % endfor
+        barrier(CLK_LOCAL_MEM_FENCE);
+
+        // Load and reduce along nupts (y direction), rows 1 to nrow
+        for (int j = 1; j < nrow; j++)
+        {
+        % for k in range(ncola):
+            idx = j*ldim + SOA_IX(i, ${k}, ${ncola});
+            r = pow(x[idx]/(atol + rtol*max(fabs(y[idx]), fabs(z[idx]))), 2);
+        % if norm == 'uniform':
+            sdata${k}[tid] = max(sdata${k}[tid], r);
+        % else:
+            sdata${k}[tid] += r;
+        % endif
+        % endfor
+            barrier(CLK_LOCAL_MEM_FENCE);
+        }
+    }
+
+    // Unrolled reduction within blocks
+    if (bid != gdim - 1)
+    {
+    % for n in range(pyfr.ilog2(sharesz)):
```

Comment:
This block is messy.  Should be a nicer way to iterate over the powers of two up to a number sharsz.